framework/src/org/checkerframework/framework/type/AnnotationClassLoader.java

```diff
@@ -0,0 +1,726 @@
+
+        // create the data structure to hold all loaded annotation classes
+        loadedAnnotations = new HashSet<Class<? extends Annotation>>();
+
+        // see if we can access the application classloader
+        ClassLoader appClassloader = checker.getClass().getClassLoader();
+        if (appClassloader != null) {
+            // if the application classloader is accessible, then directly
+            // retrieve the resource URL of the qual package
+            // resource URLs must use slashes
+            resourceURL = appClassloader.getResource(packageName.replace(DOT, SLASH));
+
+            // thread based application classloader, if needed in the future:
+            // resourceURL = Thread.currentThread().getContextClassLoader().getResource(packageName.replace(DOT, SLASH));
+        } else {
+            // if the application classloader is not accessible (which means the
+            // checker class was loaded using the bootstrap classloader)
+            // then scan the classpaths to find a jar or directory which
+            // contains the qual package and set the resource URL to that jar or
+            // qual directory
+            resourceURL = getURLFromClasspaths();
+        }
+
+        // load the annotation classes using reflective lookup
+        loadBundledAnnotationClasses();
+    }
+
+    /**
+     * Scans all classpaths and returns the resource URL to the jar which
+     * contains the checker's qual package, or the qual package directory if it
+     * exists, or null if no jar or directory contains the package
+     * 
+     * @return a URL to the jar that contains the qual package, or to the qual
+     *         package's directory, or null if no jar or directory contains the
+     *         qual package
+     */
+    private final URL getURLFromClasspaths() {
+        // Debug use, uncomment if needed to see all of the classpaths (boot
+        // classpath, extension classpath, and classpath)
+        // printPaths();
+
+        URL url = null;
+
+        // obtain all classpaths
+        Set<String> paths = getClasspaths();
+
+        // In checkers, there will be a resource URL for the qual directory. But
+        // when called in the framework (eg GeneralAnnotatedTypeFactory), there
+        // won't be a resourceURL since there isn't a qual directory
+
+        // each path from the set of classpaths will be checked to see if it
+        // contains the qual directory of a checker, if so, the the first
+        // directory or jar that contains the package will be used as the source
+        // for loading classes from the qual package
+
+        // if either a directory or a jar contains the package, resourceURL will
+        // be updated to refer to that source, otherwise resourceURL remains as null
+
+        // TODO: prefer file directory (typically in build directory) over jars?
```

Comment:
I was thinking of using the LinkedHashSet, preserving insertion order while making sure it is unique for efficiency. When I discussed this point yesterday with Prof. Dietl in person he mentioned selecting the first jar or directory in the order that is set in the command line options, providing flexibility that way. It will still scan bootclasspath before extension paths, and finally classpaths, as the classloader works according to this order.